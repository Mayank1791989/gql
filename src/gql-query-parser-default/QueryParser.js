/* @flow */
import {
  type Stream,
  type TokenState,
  type IQueryParser,
  type GQLPosition,
} from 'gql-shared/types';
import {
  LanguageParser,
  getFragmentDefinitionsAtPosition,
} from 'gql-shared/GQLParser';
import {
  genOnlineParserOptions,
  eatInterpolation,
  type InterpolationState,
} from './onlineParser';
import { type ParserOptions } from './options';
import { Source } from 'graphql';

class QueryParser extends LanguageParser implements IQueryParser {
  _opts: ParserOptions;

  constructor(options: ParserOptions) {
    super(
      genOnlineParserOptions({
        allowFragmentInterpolation: false,
        allowDocumentInterpolation: false,
        allowFragmentWithoutName: false,
        ...options,
      }),
    );
    this._opts = options;
  }

  token(stream: Stream, state: TokenState) {
    const _state: InterpolationState = (state: any);
    if (_state.interpolation) {
      const { style } = _state.interpolation;
      // NOTE: eatInterpolation mutate both stream and state
      eatInterpolation(stream, _state);
      stream._start -= 2; // to include '${' in token
      return style;
    }

    return this._parser.token(stream, state);
  }

  supportsFragmentWithoutName(): boolean {
    return this._opts.allowFragmentInterpolation || false;
  }

  getFragmentDefinitionsAtPosition(
    context: $FixMe,
    source: Source,
    position: GQLPosition,
    filterByName?: string,
  ) {
    return getFragmentDefinitionsAtPosition(
      this,
      context,
      source,
      position,
      filterByName,
    );
  }
}

export default QueryParser;
