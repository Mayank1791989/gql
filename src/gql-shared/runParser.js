/* @flow */
import MultilineCharacterStream from './MultilineCharacterStream';
import whileSafe from './whileSafe';
import printTokenState from './printTokenState';
import { type IParser, type TokenState } from './types';

export const BREAK = 'BREAK';

export default function runParser(
  parser: IParser,
  _sourceText: string,
  callback: (
    stream: MultilineCharacterStream,
    state: TokenState,
    style: string,
  ) => null | typeof BREAK,
) {
  const sourceText = _sourceText ? _sourceText : ' ';
  const state = parser.startState();
  const stream = new MultilineCharacterStream(sourceText);

  let style = '';

  let breakByCallback = false;
  whileSafe(
    {
      condition: () => !stream.eof() && !breakByCallback,
      call: () => {
        style = parser.token(stream, state);
        // eslint-disable-next-line callback-return
        const code = callback(stream, state, style);
        if (code === BREAK) {
          breakByCallback = true;
        }
        // if (style === 'js-frag') {
        // console.log(
        //   `position: ${stream.getCurrentPosition()}`,
        //   `start: ${stream.getStartOfToken()}`,
        //   `style: [${style}] token: [${stream.current()}] \n`,
        //   // 'state: ', state, '\n', printTokenState(state),
        // );
        // }
      },
      logOnInfiniteLoop: () => {
        console.log(
          `style: [${style}] token: [${stream.current()}] \n`,
          'state: ',
          state,
          '\n',
          printTokenState(state),
        );
      },
    },
    sourceText.length,
  );
}
