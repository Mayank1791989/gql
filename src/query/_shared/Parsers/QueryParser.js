/* @flow */
import { LexRules, ParseRules, isIgnored } from 'codemirror-graphql/utils/Rules';
import onlineParser from 'codemirror-graphql/utils/onlineParser';
import type { Stream, TokenState } from '../../../shared/types';
import invariant from 'invariant';

function JSInlineFragment() {
  return {
    style: 'ws-2',
    match(token) { return token.value === '${'; },
    update(state) {
      state.jsInlineFragment = { count: 1 }; // count is number of open curly braces
    },
  };
}

function eatJSInlineFragment(stream, state) {
  const { jsInlineFragment: frag } = state;
  invariant(frag, 'missing JSInlineFragment');
  stream.eatWhile((ch) => {
    if (frag.count === 0) {
      state.jsInlineFragment = null;
      return false;
    }
    if (!ch) {
      return false;
    } // eol
    if (ch === '}') {
      frag.count -= 1;
    }
    if (ch === '{') {
      frag.count += 1;
    }
    return true;
  });
}

const parserOptions = {
  eatWhitespace: stream => stream.eatWhile(ch => isIgnored(ch) || ch === ';'),
  LexRules: {
    JSInlineFragment: /\$\{/,
    ...LexRules,
  },

  ParseRules: {
    ...ParseRules,

    // relay only one definition per Relay.QL
    Document: ['Definition'],

    // only query, mutation and fragment possible in Relay.QL
    Definition(token) {
      switch (token.value) {
        case 'query': return 'Query';
        case 'mutation': return 'Mutation';
        case 'fragment': return 'FragmentDefinition';
        default: return null;
      }
    },

    Selection(token, stream) {
      if (token.value === '${') { return 'JSInlineFragment'; }
      return ParseRules.Selection(token, stream);
    },

    JSInlineFragment: [JSInlineFragment()],
  },
};

export default class QueryParser {
  _parser: any;

  constructor() {
    this._parser = onlineParser(parserOptions);
  }

  startState() {
    return this._parser.startState();
  }

  token(stream: Stream, state: TokenState) {
    if (state.jsInlineFragment) {
      eatJSInlineFragment(stream, state);
      return 'js-frag';
    }

    return this._parser.token(stream, state);
  }
}

