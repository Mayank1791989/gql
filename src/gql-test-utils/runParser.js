/* @flow */
import { type IParser } from 'gql-shared/types';
import MultilineCharacterStream from 'gql-shared/MultilineCharacterStream';

export default function runParser({
  sourceText,
  parser,
}: {
  sourceText: string,
  parser: IParser,
}) {
  const state = parser.startState();
  const stream = new MultilineCharacterStream(sourceText);
  const result = [];
  while (!stream.eof()) {
    const style = parser.token(stream, state);
    result.push({
      start: stream.getStartOfToken(),
      end: stream.getCurrentPosition(),
      text: stream.current(),
      style,
    });
  }

  return result;
}
