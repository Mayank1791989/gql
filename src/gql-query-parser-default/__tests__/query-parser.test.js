/* @flow */
import QueryParser from '../index';
import runParser from 'gql-test-utils/runParser';

test('option: allowFragmentInterpolation', () => {
  expect(
    runParser({
      sourceText: `
        fragment x on SomeType {
          name
          \${Component.getFragment('someType')}
        }
      `,
      parser: new QueryParser({
        allowFragmentInterpolation: true,
      }),
    }),
  ).toMatchSnapshot();
});

test('option: allowDocumentInterpolation', () => {
  expect(
    runParser({
      sourceText: `
        fragment x on SomeType {
          name
        }
        \${some_imported_document}
      `,
      parser: new QueryParser({
        allowDocumentInterpolation: true,
      }),
    }),
  ).toMatchSnapshot();
});
