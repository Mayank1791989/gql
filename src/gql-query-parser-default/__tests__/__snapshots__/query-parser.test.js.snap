// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`option: allowDocumentInterpolation 1`] = `
Array [
  Object {
    "end": 9,
    "start": 0,
    "style": "ws",
    "text": "
        ",
  },
  Object {
    "end": 17,
    "start": 9,
    "style": "keyword",
    "text": "fragment",
  },
  Object {
    "end": 18,
    "start": 17,
    "style": "ws",
    "text": " ",
  },
  Object {
    "end": 19,
    "start": 18,
    "style": "def",
    "text": "x",
  },
  Object {
    "end": 20,
    "start": 19,
    "style": "ws",
    "text": " ",
  },
  Object {
    "end": 22,
    "start": 20,
    "style": "keyword",
    "text": "on",
  },
  Object {
    "end": 23,
    "start": 22,
    "style": "ws",
    "text": " ",
  },
  Object {
    "end": 31,
    "start": 23,
    "style": "atom",
    "text": "SomeType",
  },
  Object {
    "end": 32,
    "start": 31,
    "style": "ws",
    "text": " ",
  },
  Object {
    "end": 33,
    "start": 32,
    "style": "punctuation",
    "text": "{",
  },
  Object {
    "end": 44,
    "start": 33,
    "style": "ws",
    "text": "
          ",
  },
  Object {
    "end": 48,
    "start": 44,
    "style": "property",
    "text": "name",
  },
  Object {
    "end": 57,
    "start": 48,
    "style": "ws",
    "text": "
        ",
  },
  Object {
    "end": 58,
    "start": 57,
    "style": "punctuation",
    "text": "}",
  },
  Object {
    "end": 67,
    "start": 58,
    "style": "ws",
    "text": "
        ",
  },
  Object {
    "end": 69,
    "start": 67,
    "style": "",
    "text": "\${",
  },
  Object {
    "end": 92,
    "start": 67,
    "style": "ws-2",
    "text": "\${some_imported_document}",
  },
  Object {
    "end": 99,
    "start": 92,
    "style": "ws",
    "text": "
      ",
  },
]
`;

exports[`option: allowFragmentInterpolation 1`] = `
Array [
  Object {
    "end": 9,
    "start": 0,
    "style": "ws",
    "text": "
        ",
  },
  Object {
    "end": 17,
    "start": 9,
    "style": "keyword",
    "text": "fragment",
  },
  Object {
    "end": 18,
    "start": 17,
    "style": "ws",
    "text": " ",
  },
  Object {
    "end": 19,
    "start": 18,
    "style": "def",
    "text": "x",
  },
  Object {
    "end": 20,
    "start": 19,
    "style": "ws",
    "text": " ",
  },
  Object {
    "end": 22,
    "start": 20,
    "style": "keyword",
    "text": "on",
  },
  Object {
    "end": 23,
    "start": 22,
    "style": "ws",
    "text": " ",
  },
  Object {
    "end": 31,
    "start": 23,
    "style": "atom",
    "text": "SomeType",
  },
  Object {
    "end": 32,
    "start": 31,
    "style": "ws",
    "text": " ",
  },
  Object {
    "end": 33,
    "start": 32,
    "style": "punctuation",
    "text": "{",
  },
  Object {
    "end": 44,
    "start": 33,
    "style": "ws",
    "text": "
          ",
  },
  Object {
    "end": 48,
    "start": 44,
    "style": "property",
    "text": "name",
  },
  Object {
    "end": 59,
    "start": 48,
    "style": "ws",
    "text": "
          ",
  },
  Object {
    "end": 61,
    "start": 59,
    "style": "",
    "text": "\${",
  },
  Object {
    "end": 95,
    "start": 59,
    "style": "js-frag",
    "text": "\${Component.getFragment('someType')}",
  },
  Object {
    "end": 104,
    "start": 95,
    "style": "ws",
    "text": "
        ",
  },
  Object {
    "end": 105,
    "start": 104,
    "style": "punctuation",
    "text": "}",
  },
  Object {
    "end": 112,
    "start": 105,
    "style": "ws",
    "text": "
      ",
  },
]
`;
